packages/context/src/binding-decorator.ts

```diff
@@ -0,0 +1,215 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/context
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {ClassDecoratorFactory, MetadataInspector} from '@loopback/metadata';
+import {Binding, BindingScope, BindingTemplate, TagMap} from './binding';
+import {Provider} from './provider';
+import {Constructor} from './value-promise';
+
+const BINDING_METADATA_KEY = 'binding.metadata';
+
+type BindingMetadata = {
+  templates: BindingTemplate[];
+  target?: Constructor<unknown>;
+};
+
+const removeNameAndKey: BindingTemplate = binding => {
+  if (binding.tagMap) {
+    delete binding.tagMap.name;
+    delete binding.tagMap.key;
+  }
+};
+
+class BindDecoratorFactory extends ClassDecoratorFactory<BindingMetadata> {
+  mergeWithInherited(inherited: BindingMetadata, target: Function) {
+    if (inherited) {
+      return {
+        templates: [
+          ...inherited.templates,
+          removeNameAndKey,
+          ...this.spec.templates,
+        ],
+        target: this.spec.target,
+      };
+    } else {
+      this.withTarget(this.spec, target);
+      return this.spec;
+    }
+  }
+
+  withTarget(spec: BindingMetadata, target: Function) {
+    spec.target = target as Constructor<unknown>;
+    return spec;
+  }
+}
+
+export type BindingScopeAndTags = {
+  scope?: BindingScope;
+  tags?: string | TagMap | (string | TagMap)[];
+};
+
+/**
+ * Decorate a class with binding configuration
+ *
+ * @example
+ * ```ts
+ * @bind((binding) => {binding.inScope(BindingScope.SINGLETON).tag('controller')}
+ * )
+ * export class MyController {
+ * }
+ * ```
+ *
+ * @param templates A list of binding scope/tags or template functions to
+ * configure the binding
+ */
+export function bind(
+  ...templates: (BindingTemplate | BindingScopeAndTags)[]
```

Comment:
Personally, I would restrict the allowed signatures to `BindingTemplate[] | BindingScopeAndTags`.

Either way, please make sure there are tests covering all different combinations allowed by this API, both for `@bind()` and `@bind.provider()`. For example: 
 - multiple templates
 - multiple scope+tags objects
 - combinations of the above

Reply:
Added tests